File: Java/src/lexer/Token.java
package lexer;

import Common.AlphabetHelper;
import Common.PeekIterator;

/**
 * Author : lihuichuan
 * Time   : 2020/11/28
 */
public class Token {
    TokenType _type;
    String _value;

    public Token(TokenType type, String value) {
        this._type = type;
        this._value = value;
    }

    public TokenType getType() {
        return _type;
    }

    public String getValue() {
        return _value;
    }

    public boolean isVariable() {
        return _type == TokenType.VARIABLE;
    }

    public boolean isScalar() {
        return _type == TokenType.INTEGER || _type == TokenType.FLOAT ||
                _type == TokenType.STRING || _type == TokenType.BOOLEAN;
    }

    /**
     * 提取变量或关键字
     *
     * @param it
     * @return Token
     */
    public static Token makeVarOrKeyword(PeekIterator<Character> it) {
        String s = "";
        while (it.hasNext()) {
            if (AlphabetHelper.isLiteral(it.peek())) {
                s += it.peek();
            } else {
                break;
            }
            it.next();
        }

        // 判断关键词OR变量
        if (Keywords.isKeyword(s)) {
            return new Token(TokenType.KEYWORD, s);
        }

        if (s.equals("true") || s.equals("false")) {
            return new Token(TokenType.BOOLEAN, s);
        }

        return new Token(TokenType.VARIABLE, s);
    }

    @Override
    public String toString() {
        return String.format("type %s, value %s", _type, _value);
    }


}

File: JS/src/lexer/Token.js
const TokenType = require("./TokenType");
const AlphabetHelper = require("./AlphabetHelper");
const LexicalException = require("./LexicalException");

const Keywords = new Set([
  "var",
  "if",
  "else",
  "for",
  "while",
  "break",
  "func",
  "return",
  "int",
  "float",
  "bool",
  "void",
  "string"
]);

class Token {
  constructor(type, value) {
    this._type = type;
    this._value = value;
  }

  getType() {
    return this._type;
  }

  getValue() {
    return this._value;
  }

  isVariable() {
    return this._type == TokenType.VARIABLE;
  }

  isValue() {
    return this.isScalar() || this.isVariable();
  }

  isType() {
    return (
      this._value === "bool" ||
      this._value === "int" ||
      this._value === "float" ||
      this._value === "void" ||
      this._value === "string"
    );
  }

  isScalar() {
    return (
      this._type == TokenType.INTEGER ||
      this._type == TokenType.FLOAT ||
      this._type == TokenType.STRING ||
      this._type == TokenType.BOOLEAN
    );
  }

  toString() {
    return `type ${this._type.type}, value ${this._value}`;
  }


  static MakeVarOrKeyword(it) {
    let s = "";

    while (it.hasNext()) {
      const c = it.peek()
      if (AlphabetHelper.isLiteral(c)) {
        s += c;
      } else {
        break;
      }
      // 不变式
      it.next();
      
    }
    if (Keywords.has(s)) {
      return new Token(TokenType.KEYWORD, s);
    }

    if(s == 'true' || s=='false'){
      return new Token(TokenType.BOOLEAN, s);
    }

    return new Token(TokenType.VARIABLE, s);

  }
}